Convert most awaitBusy calls to assertBusy

pugnascotia opened the pull request:
Closes #28450. Convert most `awaitBusy` calls to `assertBusy`, and use asserts where possible.

This PR follows on from #28548 by @liketic. However, given the time elapsed since that PR, I implemented the changes from scratch instead of trying to merge (which would probably have taken longer). The PR was still useful though - thanks!

There were a small number of places where it didn't make sense to me to call `assertBusy`, so I kept the existing calls but renamed the method to `waitUntil`. This was partly to better reflect its usage, and partly so that anyone trying to add a new call to `awaitBusy` wouldn't be able to find it.

I also didn't change the usage in `TransportStopRollupAction` as the comments state that the local `awaitBusy` method is a temporary copy-and-paste. If @polyfractal can tell me whether and how it can be removed, I'm happy to do it.

Tagging @javanna as he reviwed the original PR.

liketic committed:
Remove awaitBusy in favor of assertBusy (#28450)

liketic committed:
Merge branch 'master' into fix-issues/28450

liketic committed:
Fix comments

liketic committed:
Remove wait for relocation

liketic committed:
Merge branch 'master' into fix-issues/28450

javanna committed:
Merge branch 'master' into fix-issues/28450

liketic committed:
Remove awaitBusy in xpack

javanna committed:
Merge branch 'master' into fix-issues/28450

liketic committed:
Merge branch 'master' of github.com:elastic/elasticsearch into fix-issues/28450

liketic committed:
Remove awaitBusy

liketic committed:
Resolve conflicts

liketic committed:
Remove redundant import

Commit:
Merge branch 'master' into fix-issues/28450

Commit:
Changing a few more cases

Commit:
Merge branch 'master' into fix-issues/28450

Commit:
Changing one more awaitsBusy

liketic committed:
Merge branch 'master' of github.com:elastic/elasticsearch into fix-issues/28450

liketic committed:
Merge master

liketic committed:
Remove unused import

liketic committed:
Merge master

liketic committed:
Replace awaitBusy with assertBusy

liketic committed:
remove unused imports

liketic committed:
Ensure empty wait not throw error

elasticmachine commented:
Pinging @elastic/es-core-infra

polyfractal commented:
Let's leave the c/p of `awaitBusy` in `TransportStopRollupAction` alone for now if you don't mind.  We're using it as an actual await, not an assertion (it's waiting for the local job, on a different thread, to tick over to a new state).  We currently check the return value, and have the whole thing wrapped in a `try` too.

The comments reference a different refactor which got shelved, but I think this will be going away soon anyway because we have a different refactor in mind which would obviate the need for it :)

rjernst commented on test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java:
```diff
@@ -965,19 +965,28 @@ public long waitForDocs(final long numDocs, int maxWaitTime, TimeUnit maxWaitTim
-        while (!awaitBusy(testDocs, maxWaitTime, maxWaitTimeUnit)) {
+        while (true) {
+            try {
+                assertBusy(testDocs, maxWaitTime, maxWaitTimeUnit);
```
Perhaps we should just make the maxwaittime scale based on number of docs instead of doing several assert busy calls?

pugnascotia replied:
Happy to do this but I could use some input. The main implementation of `waitForDocs` is only called by an override, which uses a max wait time of 90s. I looked at the usages of `waitForDocs`, and the possible range for `numDocs` is 0 to 2500. (There is one test that indexes 10000 docs, but it calls `waitForDocs` multiple times). At the moment, `waitForDocs` would continue happily so long as 1 document was indexed every 90s. So the question is, how long does it take to index a document?

rjernst replied:
I think going with something conservative would be fine as a start. 200ms per doc? We can have use a hardcoded min value for the wait time to make sure it isn't too small if say we had 3 heavyweight documents being index.

pugnascotia replied:
I went with 90s due to the comment in the method's override.

rjernst replied:
90s seems like it won't give any guarantees. We will hit the test suite timeout of 20 mins long before we hit 900k seconds for 10k docs.

pugnascotia replied:
Sorry, I should have been clearer - 90s is the minimum wait time, due to the comment:

> indexing threads can wait for up to ~1m before retrying when they first try to index into a shard which is not STARTED.

Whereas I've applied a per-document time of 0.2s. So for 2500 docs, the method will wait for 0.2s * 2500 => 500s.

rjernst commented on x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/CcrIntegTestCase.java:
```diff
@@ -647,19 +647,28 @@ public long waitForDocs(final long numDocs, int maxWaitTime, TimeUnit maxWaitTim
         };
 
-        while (!awaitBusy(testDocs, maxWaitTime, maxWaitTimeUnit)) {
+        while (true) {
```
Ditto on my previous comment about adjusting max wait time.

rjernst commented on x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/AbstractUpgradeTestCase.java:
```diff
@@ -82,22 +86,30 @@ protected Settings restClientSettings() {
-                    success &= exists;
+                        .getStatusLine().getStatusCode() == 200;
+
+                    if (!exists) {
```
nit: `== false` please

pugnascotia replied:
Roger. That'll take a little getting used to :-)

rjernst reviewed (approved):
Looks good, I left some suggestions

pugnascotia commented on server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java:
```diff
@@ -787,9 +787,11 @@ public void sendRequest(Transport.Connection connection, long requestId, String
-                    } catch (InterruptedException e) {
+                        assertBusy(() -> assertFalse(
+                            "Expected there to be some initializing shards",
+                            client(blueNodeName).admin().cluster().prepareState().setLocal(true).get()
```
In this case we're asserting that this collection *isn't* empty, so if the assertion fails there's no extra information to report. That said, I tweaked the error message to be more explicit.

pugnascotia replied:
In this case we're asserting that this collection isn't empty, so if the assertion fails there's no extra information to report. That said, I tweaked the error message to be more explicit.

rjernst replied:
`not(empty())` should work. In general, using the the hamcrest matchers produces much better error messages on failures.

pugnascotia commented on server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java:
```diff
@@ -787,9 +787,11 @@ public void sendRequest(Transport.Connection connection, long requestId, String
+                            "Expected there to be some initializing shards",
+                            client(blueNodeName).admin().cluster().prepareState().setLocal(true).get()
+                                .getState().getRoutingTable().index("test").shard(0).getAllInitializingShards().isEmpty()));
+                    } catch (Exception e) {
```
So there's a couple of things here:

   1. `assertBusy` accepts a `CheckedRunnable<Exception>`, so from the compiler's point-of-view the method either needs to handle the thrown `Exception` or allow it to propagate.
   2. The method also calls `Thread.sleep()`, which throws an `InterruptedException`.

We could get around (1) by using a `CheckedRunnable<AssertionError>` (and changing how `CheckedRunnable` is defined), though that might be a problem for code that not only performs assertions but also operations that can throw other kinds of exceptions.

That still leaves (2). We could rethrow `InteruptedException` as some kind of runtime exception, but that feels horrible.

rjernst replied:
`assertBusy` will only catch (and thus retry) assertions. The inner loop looks like this:

```
try {
    codeBlock.run();
    return;
} catch (AssertionError e) {
    failures.add(e);
}
```

The reason for the `CheckedRunnable<Exception>` is to allow other errors to escape the retries. That escaping is exactly what I want to preserve here. We shouldn't keep retrying if the test hits eg an illegal state, we should let the entire test fail right then, so no need to catch Exception.

pugnascotia replied:
This particular piece of code isn't test code _per se_, rather it's a stub interface implementation, declared to only throw `IOException`. Since `assertBusy` declares it throws `Exception`, I have to handle somehow.

rjernst replied:
Since assertBusy is called from test methods, I don't think we need to handle Exception. The point is if a general Exception is thrown, the test should fail, so we just need to have the test method throw Exception.

pugnascotia replied:
When I said this isn't test code, I meant that this code is part of an implementation of the interface `StubbableTransport.SendRequestBehavior`, specifically:

```java
void sendRequest(Connection connection, long requestId, String action, TransportRequest request, TransportRequestOptions options) throws IOException;
```

So I can't take out the exception because the compiler will complain that I'm not handling `Exception`, and I can't change the method signature without also changing the interface (which I could do, if you think it's worthwhile).

pugnascotia commented on server/src/test/java/org/elasticsearch/snapshots/AbstractSnapshotIntegTestCase.java:
```diff
@@ -212,16 +212,18 @@ public static void unblockAllDataNodes(String repository) {
+                fail("No repository is blocked waiting on a data node");
+            }, timeout.millis(), TimeUnit.MILLISECONDS);
+        } catch (Exception e) {
             fail("Timeout waiting for repository block on any data node!!!");
```
As before, currently the exceptions have to be handled. I tried rewriting with `waitUntil` and just propagating the `InterruptedException`. It's slightly better:

```java
public void waitForBlockOnAnyDataNode(String repository, TimeValue timeout) throws InterruptedException {
    final boolean blocked = waitUntil(() -> {
        for (RepositoriesService repositoriesService : internalCluster().getDataNodeInstances(RepositoriesService.class)) {
            MockRepository mockRepository = (MockRepository) repositoriesService.repository(repository);
            if (mockRepository.blocked()) {
                return true;
            }
        }
        return false;
    }, timeout.millis(), TimeUnit.MILLISECONDS);

    assertTrue("No repository is blocked waiting on a data node", blocked);
}
```

What do you think?

pugnascotia commented on test/framework/src/main/java/org/elasticsearch/common/util/MockPageCacheRecycler.java:
```diff
@@ -44,9 +45,9 @@ public static void ensureAllPagesAreReleased() throws Exception {
-            if (!success) {
+            try {
+                assertBusy(() -> Sets.haveEmptyIntersection(masterCopy.keySet(), ACQUIRED_PAGES.keySet()));
+            } catch (Exception e) {
```
You've missed my deliberate mistake of omitting an assertion 😅. That said, after staring at the code again I think it should be a `waitUntil()`.

pugnascotia committed:
Merge remote-tracking branch 'liketic/fix-issues/28450' into 28450-remove-await-busy-take2

pugnascotia committed:
Replace further uses of awaitBusy with assertBusy

There were a small number of places where it didn't make sense to call
assertBusy, so I kept the existing calls but renamed awaitBusy to
waitUntil. This was partly to better reflect its usage, and partly so
that anyone trying to type "awaitBusy" wouldn't be able to find it.

pugnascotia commented:
This branch is now based off a merge of @liketic's branch into `master` i.e. it retains the original commits.

pugnascotia committed:
Rework waitForDocs to scale its timeout

Instead of calling `assertBusy` in a loop, work out a reasonable overall
timeout and await just once.

pugnascotia commented on test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java:
```diff
@@ -929,62 +929,53 @@ public ClusterHealthStatus waitForRelocation(ClusterHealthStatus status) {
+                }
+
+                // Have we managed to count enough docs yet?
+                if (lastKnownCount.get() < numDocs) {
```
The test here was `>=` before and I couldn't see why that would be correct. Please point out what I've missed.

rjernst replied:
Do we need any of this state between assertions anymore if we are no longer doing a loop? Shouldn't the assertion should be visible docs >= numDocs?

rjernst replied:
I don't think we need this state across assertions anymore? We can just do a normal query/assert num docs, no need to check for progress.

pugnascotia replied:
The JavaDoc says this about the `indexer` argument (which is optional only in the `ESIntegTestCase` implementation:

> If supplied it will be first checked for documents indexed. This saves on unneeded searches.

So it would make sense if the code checked the indexer's number of documents first, and only if it's lower that the desired count would it then perform a search (presumably hoping that more documents have been indexed?). I don't understand the code would check the indexer's number of docs, and only if that number is `>= numDocs` would it perform the search (which is just a search to find the total number of hits and doesn't actually return any documents). Unless the code needs the indexer to think it has attempted to index the desired number of docs first?

pugnascotia replied:
Aha, the indexed vs searchable difference is what I was missing. That makes much more sense now.

I had kept the document count because the method returned it, but actually none of the call sites use the number, so I've changed it to return `void` instead.

pugnascotia committed:
Sync waitForDocs implementations

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 28450-remove-await-busy

pugnascotia committed:
Simplify waitForDocs further

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 28450-remove-await-busy

pugnascotia committed:
Fix checkstyle

pugnascotia committed:
Fixes

pugnascotia committed:
Fix checkstyle again

pugnascotia committed:
checkstyle checkstyle checkstyle

pugnascotia committed:
checkstyle

pugnascotia commented on x-pack/plugin/data-frame/src/test/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameIndexerTests.java:
```diff
@@ -232,8 +232,9 @@ public void testPageSizeAdapt() throws InterruptedException {
+
             latch.countDown();
-            awaitBusy(() -> indexer.getState() == IndexerState.STOPPED);
+            assertBusy(() -> assertThat(indexer.getState(), equalTo(IndexerState.STOPPED)));
```
Pinging @elastic/ml-core - I've changed this line from `awaitBusy()` to `assertBusy()`, but since the latter requires the assertion to pass, the test now fails. I tried upping the timeout to 30s but that didn't help. Can you advise?

hendrikmuhs replied:
@benwtrent I think you are right, looks like a bug in the test.

As for explaining the test: The test uses a mock indexer which throws a circuit breaker exception on every search, the failure is handled and the indexer should change its state from `INDEXING` to `STARTED` (not `STOPPED`).

Funny this worked before. @pugnascotia can you try the suggestion? Same applies to line 251 below.

pugnascotia replied:
The suggestion change passes locally, I've pushed a fix. Thanks for the help 👍

pugnascotia committed:
Fix bug in DataFrameIndexerTests

There were previously two `awaitBusy` calls in this file whose
assertions never succeeded. However since `awaitBusy` doesn't fail if
the assertion never succeeds, the test carried on executing anyway. Now
that these calls have been changed to `assertBusy`, the test's
assertions needed to be corrected.

pugnascotia commented on server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java:
```diff
@@ -2847,13 +2847,18 @@ public void clusterStateProcessed(String source, ClusterState oldState, final Cl
-    private boolean waitForRelocationsToStart(final String index, TimeValue timeout) throws InterruptedException {
-        return awaitBusy(() -> client().admin().cluster().prepareHealth(index).execute().actionGet().getRelocatingShards() > 0,
-            timeout.millis(), TimeUnit.MILLISECONDS);
+    private void waitForRelocationsToStart(final String index, TimeValue timeout) throws Exception {
```
The called of this method never checked its return value. Now that the method is calling `assertBusy` instead of `awaitBusy`, the test fails, because the number of relocating shards is not greater than zero.

I'm not sure who to ping about this - @elastic/es-distributed maybe?

pugnascotia committed:
Test fix

pugnascotia commented:
@jakelandis Have you got time to look at a new test failure here? `git blame` tells me that you last edited the test in question. The [build scan] has the failure, but basically a test in [`ActivateWatchTests.java`](https://github.com/elastic/elasticsearch/pull/45794/files#diff-caa7fe616e139ed208e053a37de88238R95) fails because it now uses `assertBusy` instead of `awaitBusy`, and the check never succeeds.

[build scan]: https://gradle-enterprise.elastic.co/s/wkjuzgy54vwew/console-log?task=:x-pack:plugin:watcher:test#L18

pugnascotia committed:
Silence broken test

pugnascotia committed:
Fix ActivateWatchTests

pugnascotia commented:
@jakelandis never mind, I figured out what the test needs to do.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 28450-remove-await-busy

pugnascotia commented:
@elasticmachine run elasticsearch-ci/2

pugnascotia commented:
So now some tests are failing because the `security-index-template` template doesn't exist. The method `setupForTests` in `AbstractUpgradeTestCase` checks for this template.

I noticed in `xpack/security/support/SecurityIndexManager` that it used to reference `security-index-template`, and now references `security-index-template-7`. But neither of those templates are present when I run a sample failing test (`TokenBackwardsCompatibilityIT.testAccessTokensWorkInMixedCluster` for example).

@rjernst any ideas? The pertinent change is:

https://github.com/elastic/elasticsearch/pull/45794/files#diff-ed6bbaf0700db00e3a9796936dc5c831R89

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 28450-remove-await-busy

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into 28450-remove-await-busy

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into 28450-remove-await-busy

pugnascotia committed:
Correct a template name

As advised by the ML team.

pugnascotia committed:
Correct the expect templates in AbstractUpgradeTestCase

The ES Security team confirmed that they don't use templates any more,
so remove this from the expected templates. Also rewrite how the setup
code checks for templates, in order to give more information.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 28450-remove-await-busy

pugnascotia committed:
Remove an expected ML template from XPackRestTestConstants

The ML team advised that the ML tests shouldn't be waiting for any
`.ml-notifications*` templates, since such checks should happen in the
production code instead.

Also rework the template checking code in `XPackRestTestHelper` to give
more helpful failure messages.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 28450-remove-await-busy

pugnascotia commented:
@elasticmachine run elasticsearch-ci/2

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 28450-remove-await-busy

pugnascotia commented:
@elasticmachine run elasticsearch-ci/2

pugnascotia commented:
@elasticmachine run elasticsearch-ci/2

pugnascotia commented:
@elasticmachine run elasticsearch-ci/2

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into 28450-remove-await-busy

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into 28450-remove-await-busy

pugnascotia commented:
@hendrikmuhs there's no plan to backport, but I can look into that once I *finally* get this PR merged.

ywelsch commented:
I think this should definitely be backported to 7.x (I took the liberty to add the v7.5.0 label). Having the codebase diverge between master and 7.x without good reasons will complicate any other backports in the future. Perhaps the lesson learned here is to keep these PRs smaller and doing this kind of work more incrementally 😄

pugnascotia commented:
A CI that wasn't flaky would help too 😉

pugnascotia commented:
Backport PR is #47112.

pugnascotia commented:
Thanks again to @liketic for the original PR!

pugnascotia commented:
I've merged the backport to `7.x`.
